src/Covenant/JSON.hs

```diff
@@ -0,0 +1,517 @@
+
+   {tag: "EnumData"}
+   | {tag: "ProductListData"}
+   | {tag: "ConstrData"}
+   | {tag: "NewtypeData"}
+
+-}
+encodePlutusDataStrategy :: PlutusDataStrategy -> Encoding
+encodePlutusDataStrategy = error "TODO: Implement encodePlutusDataStrategy"
+
+decodePlutusDataStrategy :: Value -> ASGParser PlutusDataStrategy
+decodePlutusDataStrategy = error "TODO: Implement decodePlutusDataStrategy"
+
+{- PlutusDataConstructor encodes as a typical enumeration type:
+
+  {tag: "PlutusI"}
+  | {tag: "PlutusB"}
+  | {tag: "PlutusConstr"}
+  | {tag: "PlutusList"}
+  | {tag: PlutusMap}
+
+-}
+encodePlutusDataConstructor :: PlutusDataConstructor -> Encoding
+encodePlutusDataConstructor = error "TODO: Implement encodePlutusDataConstructor"
+
+decodePlutusDataConstructor :: Value -> ASGParser PlutusDataConstructor
+decodePlutusDataConstructor = error "TODO: Implement decodePlutusDataConstructor"
+
+{- DataDeclaration AbstractTy is a bit atypical. It is a sum type, but I think it makes more sense to encode
+   the arguments to the `DataDeclaration` constructor as an object instead of an array (to reduce the possibility for
+   frontend errors).
+
+   For example, if we have:
+
+     1. DataDeclaration "Maybe" (Count 1) [...Nothing...,...Just...] SOP
+
+   It will seralize like:
+
+     {tag: "DataDeclaration"
+     , fields: {
+        datatypeName: "Maybe",
+        datatypeBinders: 1,
+        datatypeConstructors: [...],
+        datatypeEncoding: {tag: "SOP"}
+     }}
+
+   For consistency, we do the same thing with Opaques. E.g.:
+
+     2. OpaqueData "Foo" [Plutus_I]
+
+   Will serialize to:
+
+     { tag: "OpaqueData"
+     , fields: {
+       datatypeName: "Foo",
+       opaquePlutusConstructors: {tag: "Plutus_I"}
+     }}
+-}
+encodeDataDeclarationAbstractTy :: DataDeclaration AbstractTy -> Encoding
+encodeDataDeclarationAbstractTy = error "TODO: Implement encodeDataDeclarationAbstractTy"
+
+decodeDataDeclarationAbstractTy :: Value -> ASGParser (DataDeclaration AbstractTy)
+decodeDataDeclarationAbstractTy = error "TODO: Implement decodeDataDeclarationAbstractTy"
+
+{- ASG Specific Types & Components
+
+-}
+
+{- Id encodes directly as a number. E.g.:
+
+   Id 101 -> 101
+-}
+
+encodeId :: Id -> Encoding
+encodeId = error "TODO: Implement encodeId"
+
+decodeId :: Value -> ASGParser Id
+decodeId = error "TODO: Implement decodeId"
+
+{- Ref encodes as a tagged sum type, e.g.
+
+   {tag: "AnArg", fields: [...]}
+   | {tag: "AnId": fields: [101]}
+-}
+encodeRef :: Ref -> Encoding
+encodeRef = error "TODO: Implement encodeRef"
+
+decodeRef :: Value -> ASGParser Ref
+decodeRef = error "TODO: Implement decodeRef"
+
+{- Arg encodes as an object, e.g.:
+
+   {argDebruijn: 0,
+    argIndex: 1,
+    argType: ...
+   }
+
+-}
+encodeArg :: Arg -> Encoding
+encodeArg = error "TODO: Implement encodeArg"
+
+decodeArg :: Value -> ASGParser Arg
+decodeArg = error "TODO: Implement decodeArg"
+
+{- AConstant
+
+   Serializes as a sum type without named fields:
```

Comment:
Again, module-level comment explaining how we treat sums with named fields differently to how we treat them without would help a reader a lot.